File: src/dev_health_ops/metrics/benchmarking/anomalies.py
from __future__ import annotations

import math
from collections.abc import Mapping, Sequence
from datetime import date, datetime, timedelta

from dev_health_ops.metrics.benchmarking._common import (
    MetricPoint,
    fetch_metric_series_by_scope,
    mean,
    metric_is_negative,
    population_stdev,
)
from dev_health_ops.metrics.sinks.base import BaseMetricsSink
from dev_health_ops.metrics.testops_schemas import BenchmarkAnomalyRecord


def _severity_from_z_score(z_score: float) -> str:
    magnitude = abs(z_score)
    if magnitude >= 3.0:
        return "critical"
    if magnitude >= 2.0:
        return "warning"
    return "info"


def _anomaly_direction(metric_name: str, delta: float) -> str:
    if math.isclose(delta, 0.0, abs_tol=1e-9):
        return "stable"
    improving = delta < 0 if metric_is_negative(metric_name) else delta > 0
    if improving:
        return "down" if metric_is_negative(metric_name) else "up"
    return "up" if metric_is_negative(metric_name) else "down"


def detect_metric_anomalies(
    *,
    metric_name: str,
    scope_type: str,
    series_by_scope: Mapping[str, Sequence[MetricPoint]],
    as_of_day: date,
    computed_at: datetime,
    rolling_window_days: int = 30,
    z_threshold: float = 2.0,
    volatility_threshold: float = 0.5,
    min_history_points: int = 5,
    org_id: str = "",
) -> list[BenchmarkAnomalyRecord]:
    results: list[BenchmarkAnomalyRecord] = []
    history_start = as_of_day - timedelta(days=rolling_window_days)
    for scope_key, points in sorted(series_by_scope.items()):
        ordered_points = sorted(points, key=lambda point: point.day)
        current_point = next(
            (point for point in reversed(ordered_points) if point.day <= as_of_day),
            None,
        )
        if current_point is None:
            continue
        history = [
            point.value
            for point in ordered_points
            if history_start <= point.day < current_point.day
        ]
        if len(history) < min_history_points:
            continue

        baseline_value = mean(history)
        stdev = population_stdev(history)
        if math.isclose(stdev, 0.0, abs_tol=1e-12):
            z_score = (
                0.0
                if math.isclose(current_point.value, baseline_value, abs_tol=1e-9)
                else 3.0
            )
        else:
            z_score = (current_point.value - baseline_value) / stdev

        delta = current_point.value - baseline_value
        volatility_score = 0.0
        denominator = abs(baseline_value)
        if not math.isclose(denominator, 0.0, abs_tol=1e-12):
            volatility_score = stdev / denominator

        if abs(z_score) >= z_threshold:
            improving = delta < 0 if metric_is_negative(metric_name) else delta > 0
            results.append(
                BenchmarkAnomalyRecord(
                    metric_name=metric_name,
                    scope_type=scope_type,
                    scope_key=scope_key,
                    day=current_point.day,
                    value=round(current_point.value, 4),
                    baseline_value=round(baseline_value, 4),
                    z_score=round(z_score, 4),
                    anomaly_type="improvement" if improving else "regression",
                    direction=_anomaly_direction(metric_name, delta),
                    severity=_severity_from_z_score(z_score),
                    volatility_score=round(volatility_score, 4),
                    computed_at=computed_at,
                    org_id=org_id,
                )
            )

        if volatility_score >= volatility_threshold:
            results.append(
                BenchmarkAnomalyRecord(
                    metric_name=metric_name,
                    scope_type=scope_type,
                    scope_key=scope_key,
                    day=current_point.day,
                    value=round(current_point.value, 4),
                    baseline_value=round(baseline_value, 4),
                    z_score=round(z_score, 4),
                    anomaly_type="volatility",
                    direction="volatile",
                    severity="warning" if volatility_score < 1.0 else "critical",
                    volatility_score=round(volatility_score, 4),
                    computed_at=computed_at,
                    org_id=org_id,
                )
            )
    return results


def build_metric_anomalies_from_clickhouse(
    sink: BaseMetricsSink,
    *,
    metric_name: str,
    scope_type: str,
    as_of_day: date,
    computed_at: datetime,
    rolling_window_days: int = 30,
    z_threshold: float = 2.0,
    volatility_threshold: float = 0.5,
    min_history_points: int = 5,
    org_id: str = "",
) -> list[BenchmarkAnomalyRecord]:
    start_day = as_of_day - timedelta(days=rolling_window_days + min_history_points)
    series_by_scope = fetch_metric_series_by_scope(
        sink,
        metric_name=metric_name,
        start_day=start_day,
        end_day=as_of_day,
        scope_type=scope_type,
    )
    return detect_metric_anomalies(
        metric_name=metric_name,
        scope_type=scope_type,
        series_by_scope=series_by_scope,
        as_of_day=as_of_day,
        computed_at=computed_at,
        rolling_window_days=rolling_window_days,
        z_threshold=z_threshold,
        volatility_threshold=volatility_threshold,
        min_history_points=min_history_points,
        org_id=org_id,
    )

File: internal/jobs/metrics/daily/benchmarking/anomalies.go
package benchmarking

import (
	"sort"
	"time"

	"github.com/full-chaos/dev-health-ops/internal/pythonparity"
)

// Anomaly detection defaults, mirroring detect_metric_anomalies' signature
// (anomalies.py:36-46).
const (
	DefaultRollingWindowDays = 30
	DefaultZThreshold        = 2.0
	DefaultVolatilityThresh  = 0.5
	DefaultMinHistoryPoints  = 5
)

// severityFromZScore ports anomalies.py:18-24. Boundaries are inclusive, so
// |z| exactly 3.0 is "critical" and exactly 2.0 is "warning".
func severityFromZScore(zScore float64) string {
	magnitude := abs(zScore)
	if magnitude >= 3.0 {
		return "critical"
	}
	if magnitude >= 2.0 {
		return "warning"
	}
	return "info"
}

// anomalyDirection ports anomalies.py:27-33.
//
// The isclose here is a ZERO comparison, which is provably equivalent to a
// plain |delta| <= 1e-9 at every magnitude -- pythonparity.IsCloseAbs is used
// for uniformity across the family, so a future edit that swaps the 0.0 for a
// real value cannot silently reintroduce the rel_tol bug.
func anomalyDirection(metricName string, delta float64) string {
	if pythonparity.IsCloseAbs(delta, 0.0, 1e-9) {
		return "stable"
	}
	negative := MetricIsNegative(metricName)
	improving := delta > 0
	if negative {
		improving = delta < 0
	}
	if improving {
		if negative {
			return "down"
		}
		return "up"
	}
	if negative {
		return "up"
	}
	return "down"
}

// DetectMetricAnomalies ports detect_metric_anomalies (anomalies.py:36-118).
//
// A scope can emit TWO rows for the same day: one z-score anomaly and one
// volatility anomaly. They are distinct `anomaly_type` values and the output
// table's ORDER BY includes anomaly_type, so both persist -- this is not a
// duplicate.
func DetectMetricAnomalies(
	metricName string,
	scopeType string,
	seriesByScope map[string][]MetricPoint,
	asOfDay time.Time,
	computedAt time.Time,
	rollingWindowDays int,
	zThreshold float64,
	volatilityThreshold float64,
	minHistoryPoints int,
	orgID string,
) []BenchmarkAnomalyRecord {
	historyStart := asOfDay.AddDate(0, 0, -rollingWindowDays)

	scopeKeys := make([]string, 0, len(seriesByScope))
	for scopeKey := range seriesByScope {
		scopeKeys = append(scopeKeys, scopeKey)
	}
	sort.Strings(scopeKeys)

	var results []BenchmarkAnomalyRecord
	for _, scopeKey := range scopeKeys {
		points := seriesByScope[scopeKey]
		ordered := make([]MetricPoint, len(points))
		copy(ordered, points)
		// Python sorts by day with a stable sort; sort.SliceStable keeps the
		// relative order of same-day points, which decides which one `next(...
		// reversed(...))` picks below.
		sort.SliceStable(ordered, func(i, j int) bool { return ordered[i].Day.Before(ordered[j].Day) })

		// `next((p for p in reversed(ordered) if p.day <= as_of), None)` --
		// the LAST point at or before as_of.
		var current MetricPoint
		found := false
		for index := len(ordered) - 1; index >= 0; index-- {
			if !ordered[index].Day.After(asOfDay) {
				current = ordered[index]
				found = true
				break
			}
		}
		if !found {
			continue
		}

		var history []float64
		for _, point := range ordered {
			if !point.Day.Before(historyStart) && point.Day.Before(current.Day) {
				history = append(history, point.Value)
			}
		}
		if len(history) < minHistoryPoints {
			continue
		}

		baselineValue := Mean(history)
		stdev := PopulationStdev(history)

		var zScore float64
		if pythonparity.IsCloseAbs(stdev, 0.0, 1e-12) {
			// THE REAL-VALUE isclose SITE (anomalies.py:72). Both operands are
			// metric values, routinely far greater than 1, so CPython's
			// DEFAULT rel_tol=1e-9 is active and a naive
			// math.Abs(a-b) <= 1e-9 diverges here. This branch decides 0.0 vs
			// 3.0, which selects "info" vs "critical" and whether the row is
			// emitted at all -- an alerting-level difference.
			if pythonparity.IsCloseAbs(current.Value, baselineValue, 1e-9) {
				zScore = 0.0
			} else {
				zScore = 3.0
			}
		} else {
			zScore = (current.Value - baselineValue) / stdev
		}

		delta := current.Value - baselineValue
		volatilityScore := 0.0
		denominator := abs(baselineValue)
		if !pythonparity.IsCloseAbs(denominator, 0.0, 1e-12) {
			volatilityScore = stdev / denominator
		}

		if abs(zScore) >= zThreshold {
			negative := MetricIsNegative(metricName)
			improving := delta > 0
			if negative {
				improving = delta < 0
			}
			anomalyType := "regression"
			if improving {
				anomalyType = "improvement"
			}
			results = append(results, BenchmarkAnomalyRecord{
				MetricName:      metricName,
				ScopeType:       scopeType,
				ScopeKey:        scopeKey,
				Day:             current.Day,
				Value:           round4(current.Value),
				BaselineValue:   round4(baselineValue),
				ZScore:          round4(zScore),
				AnomalyType:     anomalyType,
				Direction:       anomalyDirection(metricName, delta),
				Severity:        severityFromZScore(zScore),
				VolatilityScore: round4(volatilityScore),
				ComputedAt:      computedAt,
				OrgID:           orgID,
			})
		}

		if volatilityScore >= volatilityThreshold {
			severity := "critical"
			if volatilityScore < 1.0 {
				severity = "warning"
			}
			results = append(results, BenchmarkAnomalyRecord{
				MetricName:      metricName,
				ScopeType:       scopeType,
				ScopeKey:        scopeKey,
				Day:             current.Day,
				Value:           round4(current.Value),
				BaselineValue:   round4(baselineValue),
				ZScore:          round4(zScore),
				AnomalyType:     "volatility",
				Direction:       "volatile",
				Severity:        severity,
				VolatilityScore: round4(volatilityScore),
				ComputedAt:      computedAt,
				OrgID:           orgID,
			})
		}
	}
	return results
}
